redis/src/cluster_async/mod.rs

```diff
@@ -499,52 +529,209 @@ where
         Ok(slot_map)
     }
 
-    fn get_connection(&mut self, route: &Route) -> (String, ConnectionFuture<C>) {
-        if let Some((_, node_addrs)) = self.slots.range(&route.slot()..).next() {
-            let addr = node_addrs.slot_addr(route.slot_addr()).to_string();
-            if let Some(conn) = self.connections.get(&addr) {
-                return (addr, conn.clone());
-            }
+    async fn get_connection_from_parts(
+        addr: String,
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+    ) -> (String, ConnectionFuture<C>) {
+        if let Some(conn) = connections_and_metadata.connections.lock().await.get(&addr) {
+            return (addr, conn.clone());
+        }
 
-            // Create new connection.
-            //
-            let (_, random_conn) = get_random_connection(&self.connections, None); // TODO Only do this lookup if the first check fails
-            let connection_future = {
-                let addr = addr.clone();
-                let params = self.cluster_params.clone();
-                async move {
-                    match connect_and_check(&addr, params).await {
-                        Ok(conn) => conn,
-                        Err(_) => random_conn.await,
-                    }
+        // Create new connection.
+        //
+        let (_, random_conn) =
+            get_random_connection(&connections_and_metadata.connections, None).await; // TODO Only do this lookup if the first check fails
+        let connection_future = {
+            let addr = addr.clone();
+            let connections_and_metadata = connections_and_metadata.clone();
+            async move {
+                match connect_and_check(&addr, connections_and_metadata.cluster_params.clone())
+                    .await
+                {
+                    Ok(conn) => conn,
+                    Err(_) => random_conn.await,
                 }
             }
-            .boxed()
-            .shared();
-            self.connections
-                .insert(addr.clone(), connection_future.clone());
-            (addr, connection_future)
+        }
+        .boxed()
+        .shared();
+        connections_and_metadata
+            .connections
+            .lock()
+            .await
+            .insert(addr.clone(), connection_future.clone());
+        (addr, connection_future)
+    }
+
+    fn add_future(
+        connections_and_metadata: Arc<ConnectionsAndMetadata<C>>,
+        func: fn(C, Arc<Cmd>) -> RedisFuture<'static, Response>,
+        cmd: Arc<Cmd>,
+        futures: &mut Vec<RedisFuture<Response>>,
+        slot: &SlotAddrs,
+        slot_addr: &SlotAddr,
+    ) {
+        let addr = slot.slot_addr(slot_addr).to_string();
+        futures.push(
+            (async move {
+                let (_, conn) =
+                    Self::get_connection_from_parts(addr, connections_and_metadata).await;
+                let conn = conn.await;
+                func(conn, cmd.clone()).await
```

Comment:
where's the error handling/retries happens?

Reply:
same as any other request - in `poll_complete`